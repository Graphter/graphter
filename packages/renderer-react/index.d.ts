import ListRenderer from './list-renderer';
export { ListRenderer };
