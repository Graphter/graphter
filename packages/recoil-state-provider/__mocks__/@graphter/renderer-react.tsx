import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)