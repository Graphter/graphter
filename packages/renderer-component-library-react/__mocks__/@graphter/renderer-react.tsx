import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()