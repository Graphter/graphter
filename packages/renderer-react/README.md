# Graphter React renderer

A React framework that, when combined with node renderers such as those provided in the [Component Lib](../renderer-component-library-react/README.md) package, allows humans to manage arbitrary data structures.