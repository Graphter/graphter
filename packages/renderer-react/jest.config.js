module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts",
    "\\.(pcss|css|less)$": "identity-obj-proxy"
  },
  setupFilesAfterEnv: [ './setupTests.ts' ],
  automock: true,
  unmockedModulePathPatterns: [],
  snapshotSerializers: [
    'enzyme-to-json'
  ],
  testMatch: ['<rootDir>/**/*.test.ts*'],
  testPathIgnorePatterns: ['node_modules'],
};