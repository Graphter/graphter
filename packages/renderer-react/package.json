{
  "name": "malle-renderer-react",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "cross-env CI=1 react-scripts test --env=jsdom",
    "test:watch": "react-scripts test --env=jsdom",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "dependencies": {},
  "devDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-react": "^7.9.1",
    "@babel/preset-stage-0": "^7.8.3",
    "@svgr/rollup": "^5.3.0",
    "@types/react": "^16.9.25",
    "eslint": "^6.8.0",
    "react-scripts": "^3.4.1",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-postcss": "^2.5.0",
    "rollup-plugin-typescript2": "^0.26.0",
    "rollup-plugin-url": "^3.0.1",
    "typescript": "^3.8.3"
  }
}
