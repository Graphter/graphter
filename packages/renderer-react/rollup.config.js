import babel from 'rollup-plugin-babel'
import commonjs from 'rollup-plugin-commonjs'
import external from 'rollup-plugin-peer-deps-external'
import postcss from 'rollup-plugin-postcss'
import resolve from 'rollup-plugin-node-resolve'
import url from 'rollup-plugin-url'
import typescriptPlugin from 'rollup-plugin-typescript2'
import typescript from 'typescript';
import svgr from '@svgr/rollup'

import pkg from './package.json'

const externals = [
  ...Object.keys(pkg.peerDependencies),
  ...Object.keys(pkg.dependencies)
];
console.log('externals', externals);

export default {
  input: 'src/index.ts',
  output: [
    {
      file: pkg.main,
      format: 'cjs',
      sourcemap: true
    },
    {
      file: pkg.module,
      format: 'es',
      sourcemap: true
    }
  ],
  plugins: [
    // For debugging: { load(id) {console.log(id)} },
    external(externals),
    typescriptPlugin({
      typescript
    }),
    postcss({
      modules: true
    }),
    url(),
    svgr(),
    babel({
      exclude: 'node_modules/**',
      plugins: [ '@babel/plugin-proposal-export-default-from' ]
    }),
    resolve(),
    commonjs({
      namedExports: {
        'recoil': [ 'RecoilRoot', 'useRecoilState', 'useRecoilValue', 'useRecoilValueLoadable', 'useRecoilCallback', 'atom', 'selector' ]
      }
    })
  ]
}
