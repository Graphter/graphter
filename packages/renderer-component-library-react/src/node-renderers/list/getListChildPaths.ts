import { GetChildPathsFn, PathSegment } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";

export const getListChildPaths: GetChildPathsFn = (config, path, getNodeValue) => {
  if(!config.children?.length) throw new Error('List renderer has incorrect number of child configs')
  const childConfig = config.children[0]
  const childIds = getNodeValue<Array<string>>(path)
  const results = childIds.map((childId, i) => {
    const childPath: Array<PathSegment> = [ ...path, i ]
    const childRenderer = nodeRendererStore.get(childConfig.type)
    if(!childRenderer.getChildPaths) return [ childPath ]
    return [ childPath, ...childRenderer.getChildPaths(childConfig, childPath, getNodeValue) ]
  })
  return results.flat()
}