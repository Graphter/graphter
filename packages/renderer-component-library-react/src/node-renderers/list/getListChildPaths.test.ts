import { nodeRendererStore } from "@graphter/renderer-react";

import { getListChildPaths } from "./getListChildPaths";
import { when } from "jest-when";

const nodeRendererStoreMock = nodeRendererStore as jest.Mocked<any>

describe('getChildPaths()', () => {
  it('should return all descendent paths', () => {
    const getNodeValueMock = jest.fn()
    when(getNodeValueMock)
      .calledWith([ 'page' ])
      .mockReturnValueOnce([ 'child-id-a', 'child-id-b' ])
    const config = {
      id: 'page',
      type: 'object',
      children: [
        {
          id: 'some-child-id',
          type: 'list',
        }
      ]
    }
    when(nodeRendererStoreMock.get)
      .calledWith('list')
      .mockReturnValueOnce({
        getChildPaths: () => ([ [ 'page', 0, 'title' ] ])
      })
      .mockReturnValueOnce({
        getChildPaths: () => ([ [ 'page', 1, 'title' ] ])
      })
    const result = getListChildPaths(config, [ 'page' ], getNodeValueMock)
    expect(result).toEqual([
      [ 'page', 0 ],
      [ 'page', 0, 'title' ],
      [ 'page', 1 ],
      [ 'page', 1, 'title' ]
    ])
  })
  it.each([ null, undefined ])('should error if %o config is found for a child item', (noConfig) => {
    const getNodeValueMock = jest.fn()
    when(getNodeValueMock)
      .calledWith([ 'page' ])
      .mockReturnValueOnce([ 'child-id-a', 'child-id-b' ])
    // @ts-ignore
    expect(() => getListChildPaths(noConfig, [ 'page' ], getNodeValueMock))
      .toMatchSnapshot()
  })
  it('should skip descendent path resolution when child renderer does not implement a getChildPaths() function', () => {
    const getNodeValueMock = jest.fn()
    when(getNodeValueMock)
      .calledWith([ 'page' ])
      .mockReturnValueOnce([ 'child-id-a', 'child-id-b' ])
    const config = {
      id: 'some-config',
      type: 'string',
      children: [
        {
          id: 'some-child-id',
          type: 'string'
        }
      ]
    }
    when(nodeRendererStoreMock.get)
      .calledWith('string')
      .mockReturnValue({})
    const result = getListChildPaths(config, [ 'page' ], getNodeValueMock)
    expect(result).toEqual([
      [ 'page', 0 ],
      [ 'page', 1 ],
    ])
  })
})

export {}