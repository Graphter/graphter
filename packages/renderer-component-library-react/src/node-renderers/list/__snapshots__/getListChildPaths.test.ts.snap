// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChildPaths() should error if no config is found for a child item 1`] = `[Function]`;

exports[`getChildPaths() should error if null config is found for a child item 1`] = `[Function]`;

exports[`getChildPaths() should error if undefined config is found for a child item 1`] = `[Function]`;
